In [1]:
import pandas as pd
import numpy as np

In [2]:
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

## Analisando o conjunto de treinamento

In [3]:
#Visualizando as primeiras linhas
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
#Conferindo quantidade de linhas e colunas
train.shape

(891, 12)

In [5]:
#Adquirindo informações da quantidade de não-nulos
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [6]:
#Conferindo a quantidade de nulos nas colunas
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [7]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Tratamento dos dados de treinamento

In [9]:
#Convertendo Homem para 1 e Mulher para 0
train['Sex'] = [0 if x =='female' else 1 for x in train['Sex']]

In [10]:
#Preenchendo os valores nulos de idade com a média dos grupos gênero e classe 
train['Age'] = train['Age'].groupby([train['Sex'],train['Pclass']]).apply(lambda x: x.fillna(x.median()))

In [11]:
#Dummies - transformação de atributos categóricos em numéricos
train['Embarked_C'] = [1 if x=='C' else 0 for x in train['Embarked']]
train['Embarked_Q'] = [1 if x=='Q' else 0 for x in train['Embarked']]
train['Embarked_S'] = [1 if x=='S' else 0 for x in train['Embarked']]

## Analisando o conjunto de teste

In [12]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [13]:
test.shape

(418, 11)

In [14]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [15]:
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [16]:
test.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

## Tratamento dos dados de teste

In [17]:
#Transformando homem em 1 e mulher em 0 
test['Sex'] = [0 if x == 'female' else 1 for x in test['Sex']]

In [18]:
#Tratamento da coluna idade 
test['Age'] = test['Age'].groupby([test['Sex'],test['Pclass']]).apply(lambda x: x.fillna(x.median()))

In [19]:
#Tratamento da coluna Fare 
test['Fare'] = test['Fare'].groupby([test['Sex'],test['Pclass']]).apply(lambda x: x.fillna(x.median()))

In [20]:
#Dummies - transformação de atributos categóricos em numéricos
test['Embarked_C'] = [1 if x=='C' else 0 for x in test['Embarked']]
test['Embarked_Q'] = [1 if x=='Q' else 0 for x in test['Embarked']]
test['Embarked_S'] = [1 if x=='S' else 0 for x in test['Embarked']]

## Treinamento, validação e teste

In [21]:
from sklearn.model_selection import train_test_split

In [22]:
train['Miss'] = train['Name'].str.contains("Miss").astype(int)
train['Mrs'] = train['Name'].str.contains("Mrs").astype(int)
train['Master'] = train['Name'].str.contains("Master").astype(int)
train['Col'] = train['Name'].str.contains("Col").astype(int)
train['Major'] = train['Name'].str.contains("Major").astype(int)
train['Mr'] = train['Name'].str.contains("Mr").astype(int)

test['Miss'] = test['Name'].str.contains("Miss").astype(int)
test['Mrs'] = test['Name'].str.contains("Mrs").astype(int)
test['Master'] = test['Name'].str.contains("Master").astype(int)
test['Col'] = test['Name'].str.contains("Col").astype(int)
test['Major'] = test['Name'].str.contains("Major").astype(int)
test['Mr'] = test['Name'].str.contains("Mr").astype(int)

In [51]:
variaveis_entrada = ['Pclass','Sex','Age','SibSp','Parch','Fare','Embarked_C','Embarked_S','Miss','Mrs','Master','Col','Major','Mr']

In [52]:
X = train[variaveis_entrada]
y = train['Survived']

In [53]:
np.random.seed(0) #Mesma coisa do random_state (fixa a mesma 'ordem aleatoria')
X_treino, X_valid, y_treino, y_valid = train_test_split(X,y, test_size=0.5)

## Bibliotecas

In [41]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RepeatedKFold
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier
from lightgbm import LGBMClassifier
from sklearn.metrics import roc_auc_score
from skopt import dummy_minimize

## Validação Cruzada

In [60]:
kf = RepeatedKFold (n_splits = 3, n_repeats = 5, random_state=10)
resultados = []
for linhas_treino, linhas_valid in kf.split(X):
    #print("Treino:", linhas_treino.shape[0])
    #print("Valid:", linhas_valid.shape[0])

    X_treino, X_valid = X.iloc[linhas_treino],X.iloc[linhas_valid]
    y_treino, y_valid = y.iloc[linhas_treino],y.iloc[linhas_valid]

    #modelo = RandomForestClassifier(n_estimators = 100, n_jobs=-1, random_state=0)
    #modelo = LogisticRegression()
    #modelo = DecisionTreeClassifier()
    #modelo = MLPClassifier()
    modelo = LGBMClassifier(learning_rate=0.022,num_leaves=73,min_child_samples=4,subsample=0.69,colsample_bytree=0.47)
    modelo.fit(X_treino,y_treino)

    pp =  modelo.predict(X_valid)
    acuracia = np.mean(y_valid == pp)
    resultados.append(acuracia)

In [61]:
np.mean(resultados)

0.831425364758698

## Modelo Random Forest: 0.8069584736251402
## Modelo Regressão Logistica: 0.8228956228956231
## Modelo Decision Tree: 0.7764309764309764
## Modelo MLP: 0.8109988776655442
## Modelo LGBM sem tunar: 0.8237934904601574
## Modelo LGBM tunado: 0.831425364758698

## Hiperparâmetro modelo LGBM

In [58]:
def treinar_modelo(params):
    learning_rate = params[0]
    num_leaves = params[1]
    min_child_samples = params[2]
    subsample = params[3]
    colsample_bytree = params[4]
    
    print(params, '\n')
    
    mdl = LGBMClassifier(learning_rate=learning_rate, num_leaves=num_leaves, min_child_samples=min_child_samples,
                        subsample=subsample, colsample_bytree=colsample_bytree, random_state=0, subsample_freq=1, 
                         n_estimators=100)
    mdl.fit(X_treino, y_treino)
    
    pip = mdl.predict_proba(X_valid)[:,1]
    
    return -roc_auc_score(y_valid, pip)

space = [(1e-3, 1e-1, 'log-uniform'), #learning rate
         (2, 128), # num_leaves
         (1, 100), # min_child_samples
         (0.05, 1.0), # subsample
         (0.1, 1.0)] # colsample bytree

resultado = dummy_minimize(treinar_modelo, space, random_state=1, verbose=1, n_calls=30)

Iteration No: 1 started. Evaluating function at random point.
[0.09871192514273254, 74, 10, 0.3372159440002478, 0.23208030173540176] 

Iteration No: 1 ended. Evaluation done at random point.
Time taken: 0.0678
Function value obtained: -0.8637
Current minimum: -0.8637
Iteration No: 2 started. Evaluating function at random point.
[0.001529949829431263, 78, 72, 0.3782826906908954, 0.457090726807603] 

Iteration No: 2 ended. Evaluation done at random point.
Time taken: 0.0429
Function value obtained: -0.8183
Current minimum: -0.8637
Iteration No: 3 started. Evaluating function at random point.
[0.01195730942971637, 128, 19, 0.5483207515942279, 0.49910760440160107] 

Iteration No: 3 ended. Evaluation done at random point.
Time taken: 0.0558
Function value obtained: -0.8617
Current minimum: -0.8637
Iteration No: 4 started. Evaluating function at random point.
[0.0028784217488024557, 16, 51, 0.9182639233502714, 0.5114843271882895] 

Iteration No: 4 ended. Evaluation done at random point.
Time

In [59]:
resultado.x

[0.021930881089764206, 73, 4, 0.6880713925510277, 0.47060909107214777]

## Usando com os dados de teste

In [62]:
model = LGBMClassifier(learning_rate=0.022,num_leaves=73,min_child_samples=4,subsample=0.69,colsample_bytree=0.47)
model.fit(X,y)

prev = model.predict(test[variaveis_entrada])

In [63]:
#Criando a Serie de entrega
sub = pd.Series(prev, index = test['PassengerId'], name='Survived')

In [64]:
#Passando os dados para csv
sub.to_csv("Modelo_Oficial1.csv",header=True)